Source/WebCore/dom/ViewTransition.cpp

```diff
@@ -93,20 +93,35 @@ Ref<ViewTransition> ViewTransition::createSamePage(Document& document, RefPtr<Vi
     return viewTransition;
 }
 
-Ref<ViewTransition> ViewTransition::createInbound(Document& document, std::unique_ptr<ViewTransitionParams> params)
+// https://www.w3.org/TR/css-view-transitions-2/#resolve-inbound-cross-document-view-transition
+RefPtr<ViewTransition> ViewTransition::resolveInboundCrossDocumentViewTransition(Document& document, std::unique_ptr<ViewTransitionParams> inboundViewTransitionParams)
 {
-    Ref viewTransition = adoptRef(*new ViewTransition(document));
+    if (!inboundViewTransitionParams)
+        return nullptr;
+
+    if (document.activeViewTransition())
+        return nullptr;
+
+    auto types = document.resolveViewTransitionRule();
+    if (!types)
+        return nullptr;
+
+    RefPtr viewTransition = adoptRef(*new ViewTransition(document));
     viewTransition->suspendIfNeeded();
 
-    viewTransition->m_namedElements.swap(params->namedElements);
-    viewTransition->m_initialLargeViewportSize = params->initialLargeViewportSize;
-    viewTransition->m_initialPageZoom = params->initialPageZoom;
+    viewTransition->m_namedElements.swap(inboundViewTransitionParams->namedElements);
+    viewTransition->m_initialLargeViewportSize = inboundViewTransitionParams->initialLargeViewportSize;
+    viewTransition->m_initialPageZoom = inboundViewTransitionParams->initialPageZoom;
 
-    document.setActiveViewTransition(viewTransition.ptr());
+    document.setActiveViewTransition(RefPtr { viewTransition });
 
     viewTransition->m_updateCallbackDone.second->resolve();
     viewTransition->m_phase = ViewTransitionPhase::UpdateCallbackCalled;
 
+    viewTransition->setTypes(*types);
```

Comment:
There's this `ViewTransition` constructor that sets the type so you don't have to write a separate routine to set the type:

```
ViewTransition::ViewTransition(Document& document, RefPtr<ViewTransitionUpdateCallback>&& updateCallback, Vector<AtomString>&& initialActiveTypes)
```

But I see this constructor assumes we have an update callback, which is not needed for multi-page transition. I think this is why you wrote a new constructor that doesn't take an update callback:

```
ViewTransition::ViewTransition(Document& document)
```
which creates an empty `ViewTransitionTypeSet` (my fault, I didn't understand the justification for the new constructor, so I just put empty)

I think we should do one of the following instead
* (preferred) extend the first constructor so that if the update callback is nullptr, then we don't call the callback
* extend the second constructor to take a vector of types

Reply:
The wording of the spec is that creating a `ViewTransition` and then setting the types on it are separate steps.

Your point about us needlessly creating an empty `ViewTransitionTypeSet`, only to replace it is a good one. I'll probably just pass the types in the constructor.

Reply:
I tried to extend the first constructor in the past, but we still need to differentiate between a single-page view transition with no callback (which still wants to fire the update promise), and a cross-doc view transition, which doesn't have a callback or use the promise.